### ISSUE TITLE
JsonApiSerializer and multiple types in a collection

### ISSUE THREAD

**Issue by yuloh** [thephpleague/fractal#319]
JSON:API allows having multiple types in a collection:

```javascript
{ 
  data: [{
    type: "feed-items",
    attributes: {
     // ...
    }
  }, {
   type: "promo-feed-items",
   attributes: {
      // ...
    }
  }]
}
```

(this isn't listed in the real spec page explicitly but It's discussed [here](https://github.com/json-api/json-api/issues/862) and [here](https://gist.github.com/arches/7508c83294b623e41765))

Since we have to give a resource key when creating the collection, it's impossible to support multiple types in a collection:

```php
$resource = new Collection($books, new JsonApiBookTransformer(), 'books');
```

**Comment by neomerx** [comment]
@yuloh You're right polymorphic data are allowed by the spec. Though #253 is mostly syntax sugaring question rather than polymorphic support. The lib should somehow have access to multiple transformers in one collection and be able to work with it.

**Comment by yuloh** [comment]
I was thinking this morning that this could be worked around w/ an 'aggregate transformer' that takes other transformers as an argument and finds the correct transformer based on some test (i.e. instanceof).

**Comment by javis** [comment]
I'm needing this feature right now. @yuloh how did you solve this problem?

**Comment by yuloh** [comment]
@javis I didn't solve it, we have just avoided polymorphic collections so far because of this.

**Comment by javis** [comment]
@yuloh thanks. I'll see what I can do and then I'll update here

**Comment by ghunti** [comment]
You can check this pending PR:
https://github.com/thephpleague/fractal/pull/377

Or you can implement your own serializer, overriding the default one:
```
<?php

namespace ...

use League\Fractal\Serializer\JsonApiSerializer as BaseJsonApiSerializer;

class JsonApiSerializer extends BaseJsonApiSerializer
{
    public function item($resourceKey, array $data)
    {
        $resourceKey = $this->getTypeFromData($data);
        unset($data['type']);
        return parent::item($resourceKey, $data);
    }

    /**
     * @param array $data
     *
     * @return integer
     */
    protected function getTypeFromData(array $data)
    {
        if (!array_key_exists('type', $data)) {
            throw new InvalidArgumentException(
                'JSON API resource objects MUST have a valid type'
            );
        }
        return $data['type'];
    }
}
```

**Comment by Firtzberg** [comment]
I have managed to implement a polymorphic transformer and published it [here](https://github.com/yucadoo/polymorphic-fractal).
I needed this for a activity feed.
